cip/CIP2016-01-26-mandatory-match.adoc

```diff
@@ -0,0 +1,72 @@
+MANDATORY MATCH (u:User {id: {userId}})
+= CIP2016-01-26 - Mandatory Match
+:numbered:
+:toc:
+:toc-placement: macro
+:source-highlighter: codemirror
+
+*Author:* Stefan Plantikow <stefan.plantikow@neotechnology.com>
+
+[abstract]
+.Abstract
+--
+This CIP proposes adding a new variant of the `MATCH` clause called `MANDATORY MATCH` that fails when it does not match at least a single row in the underlying graph.
+--
+
+toc::[]
+
+== Motivation
+
+When writing applications using Cypher, many queries will start off by looking up multiple individual nodes based on an unique id with the implied expectation of matching exactly one node. When that expectation is not fulfilled (either due to sending a wrong id or the node having been deleted) it becomes important to identify which lookup failed. Achieving  this currently is not well supported by Cypher.
```

Comment:
The same can be said of non-unique matches, or of uses of clauses like UNWIND. Has there been any consideration of a general assertion mechanism? There's a lot of similarity between this concept and the concept of SKIP/LIMIT: `LIMIT n` reduces the number of rows to be [0..n]. I could imagine a variation on that which _requires_ the number of rows to be within a specified range (e.g. `REQUIRE 1`, `REQUIRE 1..`, `REQUIRE 1..4`, etc.)